webui/react/src/components/ActionDropdown.tsx

```diff
@@ -0,0 +1,93 @@
+import { Dropdown, Menu, Modal, ModalFuncProps } from 'antd';
+import { MenuInfo } from 'rc-menu/lib/interface';
+import React, { JSXElementConstructor } from 'react';
+
+import Icon from 'components/Icon';
+import handleError, { ErrorLevel, ErrorType } from 'ErrorHandler';
+import { capitalize } from 'utils/string';
+import { Eventually } from 'utils/types';
+
+import css from './ActionDropdown.module.scss';
+
+// TODO parameterize Action using Enums? https://github.com/microsoft/TypeScript/issues/30611
+export type Triggers<T extends string> = Partial<{ [key in T]: () => Eventually<void> }>
+export type Confirmations<T extends string> =
+  Partial<{ [key in T]: Omit<ModalFuncProps, 'onOk'>}>
+
+interface Props<T extends string> {
+  actionOrder: T[];
+  confirmations?: Confirmations<T>
+  onComplete?: (action?: T) => void;
+  onTrigger: Triggers<T>;
+  thing: { id: string, kind: string };
```

Comment:
**nit:** How about flattening this to `id` and `item`? The `thing` reference and grouping doesn't seem to add much.

Reply:
ditto